docs - convertTimezone, datetimeAdd, datetimeSubtract, and datetimeDiff

nllho opened the pull request:
Datetime custom expressions for 45. 

To do
- [x] Fix accepted syntax
- [x] Go through edge cases one more time

`datetimeDiff`
- [x] Test Metabase expression
- [x] Test SQL
- [x] Test Python
- [x] Check formatting and links

`convertTimezone`
- [x] Test Metabase expression
- [x] Test SQL
- [x] Test Python
- [x] Check formatting and links

nllho committed:
add now and datetimediff

nllho committed:
datetimediff in sql

jeff-bruemmer commented on docs/questions/query-builder/expressions/datetimeadd.md:
````diff
@@ -32,7 +32,7 @@ Here, **Finish By** is a custom column with the expression:
 ```
 
-You can use the [`between`](../expressions-list.md#between) or [`interval`](../expressions-list.md#interval) expressions to check if a given date falls between your start and end datetimes.
+To check if a given date falls between your start and end datetimes, you can use functions like [`between`](../expressions-list.md#between) or [`now`](../expressions/now.md).
````
Good connective tissue here.

jeff-bruemmer commented on docs/questions/query-builder/expressions/datetimeadd.md:
```diff
@@ -44,26 +44,11 @@ You can use the [`between`](../expressions-list.md#between) or [`interval`](../e
 | JSON                    | ❌                   |
 
-This table uses `timestamp` and `datetime` interchangeably---just make sure that your dates and times aren't stored as string or a number data types in your database.
+This table uses `timestamp` and `datetime` interchangeably---just make sure that your dates and times aren't stored as strings or numbers in your database.
```
We should probably link to the docs on casting to datetimes. https://www.metabase.com/docs/latest/data-modeling/metadata-editing#casting-to-a-specific-data-type

jeff-bruemmer commented on docs/questions/query-builder/expressions/datetimeadd.md:
```diff
@@ -44,26 +44,11 @@ You can use the [`between`](../expressions-list.md#between) or [`interval`](../e
-| DAK Honey Dude         | 2022-10-31 | 2022-11-14 | 2022-11-10 |
-| NO6 Full City Espresso | 2022-11-07 | 2022-11-21 | 2022-11-10 |
-| Ghost Roaster Giakanja | 2022-11-27 | 2022-12-11 |            |
+If you're using MongoDB, `datetimeAdd` will only work on versions 5 and up.
```
Worth keeping [an eye on this conversation](https://metaboat.slack.com/archives/C03SHMZ266R/p1669212168286559) to confirm driver compatibility.

jeff-bruemmer commented on docs/questions/query-builder/expressions/datetimediff.md:
```diff
@@ -0,0 +1,111 @@
+
+When you run a question using the [query_builder](https://www.metabase.com/glossary/query_builder), Metabase will convert your graphical query settings (filters, summaries, etc.) into a query, and run that query against your database to get your results.
+
+If our [cheese sample data](#calculating-an-end-date) is stored in a PostgreSQL database:
```
We should mention the `DATEDIFF` function here.

jeff-bruemmer commented on docs/questions/query-builder/expressions/datetimediff.md:
````diff
@@ -0,0 +1,111 @@
+If our [cheese sample data](#calculating-the-age-of-an-entity) is in a spreadsheet where "Aging Start" is in column B and "Aging End" is in column C:
+
+```
+DATEDIF(A22,B22,"M")
````
```suggestion
DATEDIFF(A22,B22,"M")
```
?

jeff-bruemmer commented on docs/questions/query-builder/expressions/now.md:
````diff
@@ -0,0 +1,93 @@
+now >= [Start] AND now >= [Deadline]
+```
+
+If you're looking for an adrenaline rush (and you have real-time data), you can flag the tasks that are due _right this second_:
````
> If you're looking for an adrenaline rush

Why else would you use Metabase?

jeff-bruemmer commented on docs/questions/query-builder/expressions/now.md:
```diff
@@ -0,0 +1,93 @@
+| Boolean                 | ❌                   |
+| JSON                    | ❌                   |
+
+This table uses `timestamp` and `datetime` interchangeably---just make sure that your dates and times aren't stored as strings or numbers in your database.
```
Again, link to casting docs for text/nums -> timestamps.

jeff-bruemmer commented on docs/questions/query-builder/expressions/now.md:
```diff
@@ -0,0 +1,93 @@
+
+## Limitations
+
+`now` might not actually be _now_ if you don't live in the same timezone as your Metabase report timezone.
```
Metabase, or the queried database? What about when querying a model that used `convertTimezone` to prep the data?

nllho replied:
@calherries if you could confirm the note here and my understanding below, that'd be swell :)

The spec says that `now` should return the equivalent of `convertTimezone(what_the_db_returns, reporting_timezone)`.

If your Metabase reporting timezone is EST, `now` translates to:

```sql
SELECT CURRENT TIMESTAMP AT TIME ZONE 'EST` AS now
```

Let's say your database has a column called `datetime_utc`. You can make a model with a column `datetime_pst` by writing `convertTimezone(datetime_utc, 'PST')`, which should run as:

```sql
SELECT datetime_utc AT TIME ZONE 'PST' AS datetime_pst
```

Then, if you write an expression using the model to check if `now > datetime_pst`:

```sql
SELECT CURRENT_TIMESTAMP AT TIME ZONE 'EST' > datetime_utc AT TIME ZONE 'PST'
```

The logic will check if the current date and time in EST is greater than `datetime_utc` mapped to PST.

I'll add a note that you should name custom datetime columns with the time zone if possible 😂

calherries replied:
I'm not 100% sure what you're asking here, since now translates to something slightly different for every driver. I assume this is postgres though.

> now should return the equivalent of convertTimezone(what_the_db_returns, reporting_timezone)

This isn't really true, and it's not the best way to think about it how it works. "what the db returns" ,depending on the driver, can already be in the reporting_timezone. If it's not, we convert it for the user.
It's best to think that `now` returns the current time in the reporting timezone. It's got nothing to do with `convertTimezone`.

Unfortunately, it's not quite as simple as that, because there are slight differences when the data is at rest, for example when the question is turned into a model.

If a database supports a timestamp type with time zones at rest, the result of `now()` will be a "datetime with time zone" type. Presto fits into this category.

If a database supports a timestamp type with offsets at rest, the result of `now()` will be a "zoned datetime" type. It will have an offset from UTC, not a time zone. Snowflake fits into this category.

If a database supports neither, then the result of `now()` will be a "datetime without timezone" type, and it will be stored in UTC.

But none of this matters to the user of the GUI editor. Only the SQL user might notice if they inspect the types.

The question you raise is more related to convertTimezone in general, rather than now(). And this is where users could get very confused. It's a deeply complicated topic.

The key problem with convertTimezone is that the result is converted to a type without a time zone attached. And we usually assume that datetimes without time zones are by default in the reporting_timezone, except sometimes in native SQL. There are probably some thorny edge cases and bugs we haven't considered with this.

Your example is tricky because it has a column with `datetime_utc` and a reporting timezone that is not UTC.

If it's postgres, I'm assuming `datetime_utc` is a `timestamp without time zone` column. (If it were a `timestamp with time zone` column it couldn't be in UTC, because the time zone at rest is the database timezone). So in order to create a datetime_pst column, yes, the user should do `convertTimezone(datetime_utc, 'PST', 'UTC')`. The 'UTC' argument needs to be given because the `datetime_utc` column doesn't have a time zone. So leaving it out would make us assume the column's time zone is implicitly the reporting time zone. Then, the result of convertTimezone would be another column of type `timestamp without time zone`.

If you did a comparison like `now < datetime_pst` where datetime_pst is a model column, that would translate into:
```
SELECT NOW() > datetime_pst;
```

The problem is `datetime_pst` is a timestamp without time zone, and in comparisons with `>` timestamp without time zones are assumed to be in the database time zone. If we were writing native SQL, we would want to write 
```
SELECT NOW() AT TIME ZONE 'PST'  > datetime_pst;
```
One way to write this with a custom expression is `convertTimezone(now, "PST") < datetime_pst`.

An alternative is to convert `datetime_pst` into the reporting_timezone before comparison:
```
now < convertTimezone(datetime_pst, 'EST', 'PST')
```

So, trying to summarize:
- `now`'s result implicitly takes the time zone of the reporting timezone if set
- `convertTimezone`'s result is a type without a time zone
- If you want to do comparisons between a timestamp with time zone and timestamp without time zone, you should convert them to the same timezone first.

So that's some of what users need to know. I'm sorry I didn't have enough time to write a shorter explanation.

calherries replied:
cc @qnkhuat If you have anything to add on convertTimezone

nllho replied:
@calherries This is incredibly helpful, thank you for writing it up 🙏

jeff-bruemmer reviewed (commented):
Looking good so far. Thanks for the quick turnaround.

nllho committed:
jeff's feedback

nllho committed:
converttimezone wip

nllho committed:
converttimezone draft

nllho committed:
nits

qnkhuat commented on docs/questions/query-builder/expressions/converttimezone.md:
```diff
@@ -0,0 +1,139 @@
+title: ConvertTimezone
+---
+
+# ConvertTimezone
```
I'm not very good at wording but here is a diagram of how we process convertTimezone depending on the input.

![Screenshot 2022-11-25 at 17 36 48](https://user-images.githubusercontent.com/25661381/203964971-89eb8539-c500-48a1-ba04-b463c1e71125.png)

qnkhuat replied:
[convert_timezone.zip](https://github.com/metabase/metabase/files/10091711/convert_timezone.zip)
just drop this into draw.io if you want to view the diagram

qnkhuat commented on docs/questions/query-builder/expressions/converttimezone.md:
```diff
@@ -0,0 +1,139 @@
+| Syntax                                                                                                    | Example                                                       |
+|-----------------------------------------------------------------------------------------------------------|---------------------------------------------------------------|
+| `convertTimezone(column, target, source)`                                                                 | `convertTimezone("December 28, 2022, 12:00:00", "EST, "PST")` |
+| Assigns the target time zone to your datetime column. You can optionally specify a source time zone too.  | `December 28, 2022, 9:00:00`                                  |
```
what do you mean by assign? 
My understanding of assign is like appending to a string without changing anything.

If that's the case then we don't simply just assign the timezone, but we actually `shift` the time so that the input column is displayed in the `target-timezone`

qnkhuat commented on docs/questions/query-builder/expressions/converttimezone.md:
````diff
@@ -0,0 +1,139 @@
+If our [datetime sample data](#creating-custom-report-dates) is stored in a PostgreSQL database:
+
+```sql
+SELECT original_time::TIMESTAMP AT TIME ZONE 'EST' AS team_report_time_est
````
it's not clear to me what is the type of `original_time` is it a timestamptz or timestamp type?

qnkhuat commented on docs/questions/query-builder/expressions/converttimezone.md:
```diff
@@ -0,0 +1,139 @@
+| ----------------------- | -------------------- |
+| String                  | ❌                   |
+| Number                  | ❌                   |
+| Timestamp               | ✅                   |
```
can we make a distinciton between `timestamp with time zone` and `timestamp without time zone`?

nllho replied:
This table is the same across all the custom expression docs. It's just there to help people do a quick check for common data type errors, so won't expand it into more detail than this (otherwise we could make a case to split out the other data types too!).

I'll add a note that the function will error if you provide a `source` parameter on `timestamp with time zone`.

nllho committed:
wording

nllho committed:
add info about source time zones

nllho committed:
now - tested sql, tested python, reviewed formatting

nllho committed:
datetimediff - tested sql, tested python, reviewed formatting

nllho committed:
converttimezone - tested metabase, sql, and python examples

nllho committed:
datetimediff - update example

nllho committed:
converttimezone - update examples

nllho committed:
converttimezone - formatting

nllho committed:
links

nllho committed:
links actually

nllho committed:
update expressions-list

nllho commented:
@jeff-bruemmer I ended up putting everything in one PR because it was easier to keep track of feedback (especially when thinking about the way that these functions will interact). Let me know if you'd prefer to split them up once we've gone through review.

Since the last review, I added `convertTimezone` (which probably deserves the closest look because time zones suck), updated `expressions-list`, and made some updates to `now` and `datetimeDiff` based on feedback.

jeff-bruemmer commented on docs/questions/query-builder/expressions/datetimediff.md:
```diff
@@ -0,0 +1,121 @@
+
+## Parameters
+
+Units can be any of: "year", "quarter", "month", "day", "hour", "second", or "millisecond".
```
Maybe a list here?

jeff-bruemmer commented on docs/questions/query-builder/expressions/datetimediff.md:
````diff
@@ -0,0 +1,121 @@
+datetimeDiff([Aging Start], [Aging End], "month")
+```
+
+Yes, it looks a bit wrong, but the spreadsheet function really is `DATEDIF()`, not `DATEDIFF()`.
````
```suggestion
Yes, `DATEDIF` looks a bit wrong, but the spreadsheet function really is `DATEDIF()` with one "f", not `DATEDIFF()`.
```

jeff-bruemmer commented on docs/questions/query-builder/expressions/converttimezone.md:
```diff
@@ -0,0 +1,155 @@
+
+## Parameters
+
+- `target` is the time zone you want to assign to your column.
+- The name of the `target` time zone depends on your database. For example, you may have to use "Canada/Eastern" instead of "EST".
```
Should be a single list time.

nllho committed:
link

jeff-bruemmer commented on docs/questions/query-builder/expressions/converttimezone.md:
```diff
@@ -0,0 +1,155 @@
+
+- The inner `convertTimezone` converts your column to UTC (basically, you're setting your own "source" time zone).
+- The outer `convertTimezone` converts the UTC output to your actual target time zone (EST).
+- You must provide the `source` parameter in the outer expression (because the output of the inner `convertTimezone` expression is a timestamp without time zone metadata).
```
As I said above, we should be explicit about this timezoneless return type in the definition above (plus it's good to reiterate here as well).

jeff-bruemmer commented on docs/questions/query-builder/expressions/converttimezone.md:
````diff
@@ -0,0 +1,155 @@
+If you're unsure about the time zone for **Source Time**, you should use two `convertTimezone` expressions to create **Team Report Time (EST)** :
+
+```
+convertTimezone(convertTimezone([Source Time], 'UTC'), 'EST', 'UTC')
````
Can we provide an explanatory counterfactual here to explain why someone would do this? That is: what would/could happen if we didn't know the source time zone, and we didn't nest `convertTimezone` expressions to convert the original to UTC?

(Aside: this baselining to UTC seems like something `covertTimezone` should do for you.)

nllho replied:
Gonna remove this suggestion. I double checked the logic for `convertTimezone` and I think I misremembered this comment from Cal's explanation (should go in the `now` doc instead):

> If you want to do comparisons between a timestamp with time zone and timestamp without time zone, you should convert them to the same time zone first.

jeff-bruemmer reviewed (approved):
Nice work here, @nllho. I'm approving to unblock you, but let me know if you have any questions re: my comments. You may want to get someone who worked on `convertTimezone` to vet your entry on it, as they might be able to point out some potential land mines.

nllho commented on docs/questions/query-builder/expressions/converttimezone.md:
```diff
@@ -0,0 +1,155 @@
+- **Source Time** is stored _with_ time zone metadata in your database, and
+- your Metabase displays datetimes or timestamps in your database's time zone.
+
+See [Limitations](#limitations) if you're interested in more gory details.
```
Definitely.

nllho committed:
jeff's feedback

nllho committed:
converttimezone - more time zone info

nllho committed:
converttimezone - update related functions

nllho committed:
syntax updates

nllho committed:
links

nllho commented on docs/questions/query-builder/expressions/converttimezone.md:
```diff
@@ -0,0 +1,228 @@
+title: ConvertTimezone
+---
+
+# ConvertTimezone
```
@qnkhuat @calherries 

Could you please take another look at `convertTimezone`? I've made a lot of updates since last time. I think this doc is the most complex, so it should be a good catch-all for timestamp nuances in general (I'll follow up if I need to generalize the feedback to any other datetime docs).

By end of week would be amazing, but I think these docs can wait until next week too. Thank you!

qnkhuat replied:
In general, I think this is much clearer. only have some nitty comments :)

nllho commented:
@jeff-bruemmer Do you mind taking one more pass at `convertTimezone` tomorrow? I made quite a few updates to clear things up further :)

qnkhuat commented on docs/questions/query-builder/expressions/converttimezone.md:
```diff
@@ -0,0 +1,228 @@
+- The name of the time zone you want to assign to your column.
+- Time zone names are database-dependent (such as "Canada/Eastern" vs. "EST").
+
+`source`:
```
should we have a description for this like we have for `target` too?

qnkhuat commented on docs/questions/query-builder/expressions/converttimezone.md:
```diff
@@ -0,0 +1,228 @@
+
+`target`:
+- The name of the time zone you want to assign to your column.
+- Time zone names are database-dependent (such as "Canada/Eastern" vs. "EST").
```
To be precise but we use the `tz database`.https://en.wikipedia.org/wiki/List_of_tz_database_time_zones

This is what most DBs use, only sqlserver accepts WindowTimezone which is a bit different (`Pacific Standard Time` vs `PST`)

Do you think we should have this information included?

qnkhuat commented on docs/questions/query-builder/expressions/converttimezone.md:
````diff
@@ -0,0 +1,228 @@
+convertTimezone([Source Time], 'EST')
+```
+
+If **Source Time** is stored as a `timestamp without time zone`, you _must_ provide the `source` time zone as 'UTC':
````
```suggestion
If **Source Time** is stored as a `timestamp without time zone`, you _must_ provide the `source` time zone ( in this case it is "UTC" ):
```

I feel like the combination of `must`  + a specific timezone could give the wrong idea that it has to be `UTC` all the time?

nllho replied:
@qnkhuat Yes, I remember being confused about this! I wasn't sure if `timestamp without time zone` would get treated as UTC by default.

What about this:
> If **Source Time** is stored as a `timestamp without time zone`, you _must_ provide the `source` time zone (which will depend on your database time zone):

qnkhuat commented on docs/questions/query-builder/expressions/datetimeadd.md:
```diff
@@ -6,15 +6,30 @@ title: DatetimeAdd
+- "millisecond"
+
+`amount`:
+- A whole number.
```
it works with real number too, `+ 3.5 hours`

nllho committed:
ngoc's feedback

nllho committed:
links

jeff-bruemmer commented on docs/questions/query-builder/expressions/now.md:
```diff
@@ -0,0 +1,111 @@
+title: Now
+---
+
+# Now
```
@nllho Unfortunately, `now` has been pushed to the next release. I would excise this content from this PR, and ship it (without backporting it) in a separate PR.

nllho committed:
remove now and references to now

jeff-bruemmer commented on docs/questions/query-builder/expressions/converttimezone.md:
```diff
@@ -0,0 +1,230 @@
+| `convertTimezone(column, target, source)`                             | `convertTimezone("2022-12-28T12:00:00", "Canada/Pacific", "Canada/Eastern")`      |
+| Shifts a timestamp from the source time zone to the target time zone. | Returns the value `2022-12-28T09:00:00`, displayed as `December 28, 2022 9:00 AM` |
+
+Timestamps and time zones are rather nasty to work with (it's easy to make mistakes, and difficult to catch them), so you should only try to use `convertTimezone` if the interpretation of your data is very sensitive to time-based cutoffs.
```
```suggestion
Timestamps and time zones are rather nasty to work with (it's easy to make mistakes, and difficult to catch them), so you should only try to use `convertTimezone` if the interpretation of your data is sensitive to time-based cutoffs.
```

jeff-bruemmer commented on docs/questions/query-builder/expressions/converttimezone.md:
```diff
@@ -0,0 +1,230 @@
+| Boolean                 | ❌                   |
+| JSON                    | ❌                   |
+
+If your timestamps are stored as strings or numbers in your database, you can [cast them to timestamps](../../../data-modeling/metadata-editing.md#casting-to-a-specific-data-type) from the Data Model page.
```
```suggestion
If your timestamps are stored as strings or numbers in your database, an admin can [cast them to timestamps](../../../data-modeling/metadata-editing.md#casting-to-a-specific-data-type) from the Data Model page.
```

jeff-bruemmer commented on docs/questions/query-builder/expressions/datetimesubtract.md:
```diff
@@ -44,13 +59,13 @@ You can use the [`between`](../expressions-list.md#between) or [`interval`](../e
 | JSON                    | ❌                   |
 
-This table uses `timestamp` and `datetime` interchangeably---just make sure that your dates and times aren't stored as string or a number data types in your database.
+This table uses `timestamp` and `datetime` interchangeably. If your dates and times are stored as strings or numbers in your database, you can [cast them to datetimes](../../../data-modeling/metadata-editing.md#casting-to-a-specific-data-type) from the Data Model page.
```
```suggestion
This table uses `timestamp` and `datetime` interchangeably. If your dates and times are stored as strings or numbers in your database, an admin can [cast them to datetimes](../../../data-modeling/metadata-editing.md#casting-to-a-specific-data-type) from the Data Model page.
```

jeff-bruemmer reviewed (commented):
Nice work, @nllho!

nllho committed:
jeff's feedback

nllho committed:
nits

qnkhuat committed:
Merge branch 'master' into docs-datetime-functions

qnkhuat committed:
update convert-timezone helper text

nllho committed:
update expressions-list

nllho committed:
move today workaround out of limitations

nllho committed:
Merge branch 'docs-datetime-functions' of https://github.com/metabase/metabase into docs-datetime-functions

nllho committed:
nits

nllho committed:
update date formats

nllho committed:
Merge branch 'master' into docs-datetime-functions

nllho commented on frontend/src/metabase-lib/expressions/helper-text-strings.ts:
```diff
@@ -791,7 +793,7 @@ const helperTextStrings: HelpText[] = [
         name: t`source`,
-        description: t`(Optional) The current timezone of your column. The default timezone is your report timezone.`,
+        description: t`The current time zone. Only required for timestamps with no time zone.`,
       },
```
@mazameli Any issues with these helper text updates (conflicts with translation)?
